hardware/MQTT.cpp

```diff
@@ -945,3 +979,796 @@ void MQTT::SendSceneInfo(const uint64_t SceneIdx, const std::string & /*SceneNam
+			if (ret)
+			{
+				bIsJSON = root.isObject();
+			}
+			std::string szValue;
+			if (bIsJSON)
+			{
+				std::string value_template = pSensor->value_template;
+				if (value_template.empty())
+					continue; //we need a value template (could be "{\"state\":\"OFF\",\"brightness\":255,\"color\":{\"r\":255,\"g\":255,\"b\":255}}")
+
+				Json::Value key_value = root;
+
+				if (value_template.find('.') != std::string::npos)
+				{
+					std::vector<std::string> strarray;
+					StringSplit(value_template, ".", strarray);
+					if (strarray.size()>1)
+					{
+						for (auto ittKey : strarray)
+						{
+							if (!key_value[ittKey].empty())
+							{
+								if (key_value[ittKey].isObject())
+									key_value = key_value[ittKey];
+								else
+									value_template = ittKey;
+							}
+							else
+							{
+								//key not found!
+								continue;
+							}
+						}
+					}
+				}
+				
+
+				if (key_value[value_template].empty())
+					continue; // key not found
+				szValue = key_value[value_template].asString();
+			}
+			else
+			{
+				szValue = qMessage;
+			}
+			pSensor->last_value = szValue;
+			pSensor->last_received = mytime(nullptr);
+#ifdef _DEBUG
+			std::string szLogMessage = std_format("%s (value: %s", pSensor->name.c_str(), pSensor->last_value.c_str());
+			if (!pSensor->unit_of_measurement.empty())
+			{
+				szLogMessage += " " + pSensor->unit_of_measurement;
+			}
+			szLogMessage += ")";
+			Log(LOG_NORM, "MQTT received: %s", szLogMessage.c_str());
+#endif
+			if (pSensor->component_type == "sensor")
+				handle_auto_discovery_sensor(pSensor, message->retain);
+			else if (pSensor->component_type == "switch")
+				handle_auto_discovery_switch(pSensor, message->retain);
+			else if (pSensor->component_type == "binary_sensor")
+				handle_auto_discovery_binary_sensor(pSensor, message->retain);
+		}
+		else if (pSensor->availability_topic == topic)
+		{
+			handle_auto_discovery_availability(pSensor, qMessage, message->retain);
+		}
+	}
+}
+
+uint64_t MQTT::UpdateValueInt(int HardwareID, const char *ID, unsigned char unit, unsigned char devType, unsigned char subType, unsigned char signallevel, unsigned char batterylevel, int nValue,
+			      const char *sValue, std::string &devname, bool bUseOnOffAction)
+{
+	uint64_t DeviceRowIdx = m_sql.UpdateValue(HardwareID, ID, unit, devType, subType, signallevel, batterylevel, nValue, sValue, devname, bUseOnOffAction);
+	if (DeviceRowIdx == (uint64_t)-1)
+		return -1;
+	if (m_bOutputLog)
+	{
+		std::string szLogString = RFX_Type_Desc(devType, 1) + std::string("/") + std::string(RFX_Type_SubType_Desc(devType, subType)) + " (" + devname + ")";
+		Log(LOG_NORM, szLogString);
+	}
+	m_mainworker.sOnDeviceReceived(m_HwdID, DeviceRowIdx, devname, nullptr);
+	m_notifications.CheckAndHandleNotification(DeviceRowIdx, m_HwdID, ID, devname, unit, devType, subType, nValue, sValue);
+	m_mainworker.CheckSceneCode(DeviceRowIdx, devType, subType, nValue, sValue, "MQTT Auto");
```

Comment:
Suggestion: shouldn't these 3 _signaling_ functions be moved to a single call in the mainworker? That way, the hardware module does not need to have any knowledge about which different _signaling_ functions it has to call.

Can be used by other Hardware modules as well.

And when making updates to any _signaling_ within Domoticz, it can be done in the mainworker and no need to update the different hardware modules.

Reply:
True, but would mean some good investigation... not part of this scope ;) ... and certainly will take some time to investigate and change...
I made a function for this in this class, that does the trick quite well at the moment